File: C++/CountingSort.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <algorithm>
#include <deque>
#include <queue>
#include <stack>
#include <set>
#include <string>
#include <string.h>
#include <array>
#include <unordered_map>
#include <unordered_set>
#include <iomanip>
using namespace std;
typedef long long ll;
typedef long double ld;
#define pb push_back

// O(n + maxVal)
void countingSort(int n, int arr[], int maxVal) {
    int output[n];
    int count[maxVal+1];
    
    for (int i = 0; i < n; i++) count[arr[i]]++;
    for (int i = 1; i <= maxVal; i++) count[i] += count[i-1];
    for (int i = n-1; i >= 0; i--) {
        output[count[arr[i]] - 1] = arr[i];
        count[arr[i]]--;
    }
    for (int i = 0; i < n; i++) arr[i] = output[i];
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    
}

File: Java/CountingSort.java

import java.util.*;
import java.io.*;

public class CountingSort {

	public static void main(String[] args) {

	}
	
	// O(n + maxVal)
	public static void countingSort(int[] arr, int maxVal) {
		int n = arr.length;
		int[] sortedArr = new int[n];
		int[] count = new int[maxVal + 1];
		
		for (int i = 0; i < n; i++) {
			count[arr[i]]++;
		}
		for (int i = 1; i <= maxVal; i++) {
			count[i] += count[i-1];
		}
		for (int i = n-1; i >= 0; i--) {
			sortedArr[count[arr[i]] - 1] = arr[i];
			count[arr[i]]--;
		}
		for (int i = 0; i < n; i++) {
			arr[i] = sortedArr[i];
		}
	}
}